integrations/terraform_cloud/client.py

```diff
@@ -0,0 +1,185 @@
+import httpx
+from typing import Any, AsyncGenerator, Optional
+from loguru import logger
+
+
+TERRAFORM_WEBHOOK_EVENTS = [
+    "run:applying",
+    "run:completed",
+    "run:created",
+    "run:errored",
+    "run:needs_attention",
+    "run:planning",
+]
+
+
+PAGE_SIZE = 100
+
+
+class TerraformClient:
+    def __init__(self, terraform_base_url: str, auth_token: str) -> None:
+        self.terraform_base_url = terraform_base_url
+        self.base_headers = {
+            "Authorization": f"Bearer {auth_token}",
+            "Content-Type": "application/vnd.api+json",
+        }
+        self.api_url = f"{self.terraform_base_url}/api/v2"
+        self.client = httpx.AsyncClient(headers=self.base_headers)
+
+    async def send_api_request(
+        self,
+        endpoint: str,
+        method: str = "GET",
+        query_params: Optional[dict[str, Any]] = None,
+        json_data: Optional[dict[str, Any]] = None,
+    ) -> dict[str, Any]:
+        logger.info(f"Requesting Terraform Cloud data for endpoint: {endpoint}")
+        try:
+            response = await self.client.request(
+                method=method,
+                url=f"{self.api_url}/{endpoint}",
+                params=query_params,
+                json=json_data,
+            )
+            response.raise_for_status()
+
+            logger.info(
+                f"Successfully retrieved Terraform Cloud data for endpoint: {endpoint}"
+            )
+
+            return response.json()
+
+        except httpx.HTTPStatusError as e:
+            logger.error(
+                f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+            )
+            raise
+
+        except httpx.HTTPError:
+            logger.exception("An error occured while fetching terraform cloud data")
+            raise
+
+    async def get_organizations(self) -> list[dict[str, Any]]:
+        response = await self.send_api_request(endpoint="organizations")
+        return response.get("data", [])
+
+    async def get_single_workspace(self, workspace_id: str) -> dict[str, Any]:
+        response = await self.send_api_request(endpoint=f"workspaces/{workspace_id}")
+        return response.get("data", {})
+
+    async def get_single_run(self, run_id: str) -> dict[str, Any]:
+        response = await self.send_api_request(endpoint=f"runs/{run_id}")
+        return response.get("data", {})
+
+    async def get_state_version_output(self, state_version_id: str) -> dict[str, Any]:
+        response = await self.send_api_request(
+            endpoint=f"state-versions/{state_version_id}/outputs"
+        )
+        return response.get("data", {})
+
+    async def get_paginated_workspaces(

```

Comment:
You might try to get workspaces twice at the same resync
Once for runs and once for workspaces

Please cache it